History archivals fail if a history item's hid is null

**Describe the bug**
When using tasked history export archives, if an item in that history has an empty (`null`/`None`) `hid`, exporting fails:

```pytb
Traceback (most recent call last):
  File "/cvmfs/main.galaxyproject.org/venv/lib/python3.11/site-packages/celery/app/trace.py", line 477, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/cvmfs/main.galaxyproject.org/venv/lib/python3.11/site-packages/celery/app/trace.py", line 760, in __protected_call__
    return self.run(*args, **kwargs)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/celery/__init__.py", line 184, in wrapper
    rval = app.magic_partial(func)(*args, **kwds)
  File "/cvmfs/main.galaxyproject.org/venv/lib/python3.11/site-packages/lagom/wrapping.py", line 28, in _bound_func
    return inner_func(*bound_args, **bound_kwargs)
  File "/cvmfs/main.galaxyproject.org/venv/lib/python3.11/site-packages/lagom/wrapping.py", line 45, in _error_handling_func
    return func(*args, **kwargs)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/celery/tasks.py", line 400, in write_history_to
    model_store_manager.write_history_to(request)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/managers/model_stores.py", line 217, in write_history_to
    with model.store.get_export_store_factory(
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/store/__init__.py", line 2462, in __exit__
    self._finalize()
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/store/__init__.py", line 2830, in _finalize
    super()._finalize()
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/store/__init__.py", line 2331, in _finalize
    datasets_attrs_out.write(to_json(datasets_attrs))
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/store/__init__.py", line 2327, in to_json
    return json_encoder.encode([a.serialize(self.security, self.serialization_options) for a in attributes])
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/store/__init__.py", line 2327, in <listcomp>
    return json_encoder.encode([a.serialize(self.security, self.serialization_options) for a in attributes])
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 399, in serialize
    return self._serialize(id_encoder, serialization_options)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 5272, in _serialize
    self._handle_serialize_files(id_encoder, serialization_options, rval)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 4983, in _handle_serialize_files
    serialization_options.serialize_files(self, rval)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/__init__.py", line 387, in serialize_files
    self.serialize_files_handler.serialize_files(dataset, as_dict)
  File "/cvmfs/main.galaxyproject.org/galaxy/lib/galaxy/model/store/__init__.py", line 1976, in serialize_files
    assert dataset_hid, as_dict
AssertionError: {'model_class': 'HistoryDatasetAssociation', 'create_time': '2010-03-08 01:04:31.488817', 'update_time': '2010-03-08 01:04:40.110421', 'name': 'Convert BED to GeneTrack Index on data 27', 'info': '', 'blurb': 'data', 'peek': 'binary data', 'extension': 'genetrack', 'metadata': {'dbkey': 'hg18'}, 'designation': 'output1', 'deleted': False, 'visible': True, 'dataset_uuid': None, 'validated_state': 'unknown', 'validated_state_message': None, 'encoded_id': 'deadbeef', 'state': 'ok', 'hid': None, 'annotation': None, 'tags': [], 'tool_version': None, 'history_encoded_id': 'deadbeef', 'copied_from_history_dataset_association_id_chain': []}
```

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 24.0.1.dev0
Commit: 5addd0093380118954d73ead601cc5ecc0e8b154

--- a/lib/galaxy/model/store/__init__.py
+++ b/lib/galaxy/model/store/__init__.py
@@ -2006,7 +2006,7 @@ class DirectoryModelExportStore(ModelExportStore):
 
             if len(file_list):
                 extra_files_target_filename = get_export_dataset_extra_files_dir_name(
-                    as_dict["name"], as_dict["extension"], as_dict["encoded_id"], conversion_key=conversion_key
+                    as_dict["encoded_id"], conversion_key=conversion_key
                 )
                 arcname = os.path.join(dir_name, extra_files_target_filename)
                 add(extra_files_path, os.path.join(export_directory, arcname))
@@ -2982,7 +2982,7 @@ def get_export_dataset_filename(name: str, ext: str, encoded_id: str, conversion
         return f"{base}_{encoded_id}_conversion_{conversion_key}.{ext}"
 
 
-def get_export_dataset_extra_files_dir_name(name: str, ext: str, encoded_id: str, conversion_key: Optional[str]) -> str:
+def get_export_dataset_extra_files_dir_name(encoded_id: str, conversion_key: Optional[str]) -> str:
     if not conversion_key:
         return f"extra_files_path_{encoded_id}"
     else:

--- a/lib/galaxy/model/store/__init__.py
+++ b/lib/galaxy/model/store/__init__.py
@@ -1972,8 +1972,6 @@ class DirectoryModelExportStore(ModelExportStore):
 
         dir_name = "datasets"
         dir_path = os.path.join(export_directory, dir_name)
-        dataset_hid = as_dict["hid"]
-        assert dataset_hid, as_dict
 
         if dataset.dataset.id in self.dataset_id_to_path:
             file_name, extra_files_path = self.dataset_id_to_path[dataset.dataset.id]
@@ -1992,7 +1990,7 @@ class DirectoryModelExportStore(ModelExportStore):
                 self.serialization_options.get_identifier(self.security, conversion) if conversion else None
             )
             target_filename = get_export_dataset_filename(
-                as_dict["name"], as_dict["extension"], dataset_hid, conversion_key=conversion_key
+                as_dict["name"], as_dict["extension"], as_dict["encoded_id"], conversion_key=conversion_key
             )
             arcname = os.path.join(dir_name, target_filename)
             src = file_name
@@ -2008,7 +2006,7 @@ class DirectoryModelExportStore(ModelExportStore):
 
             if len(file_list):
                 extra_files_target_filename = get_export_dataset_extra_files_dir_name(
-                    as_dict["name"], as_dict["extension"], dataset_hid, conversion_key=conversion_key
+                    as_dict["name"], as_dict["extension"], as_dict["encoded_id"], conversion_key=conversion_key
                 )
                 arcname = os.path.join(dir_name, extra_files_target_filename)
                 add(extra_files_path, os.path.join(export_directory, arcname))
@@ -2973,22 +2971,22 @@ def tar_export_directory(export_directory: StrPath, out_file: StrPath, gzip: boo
             store_archive.add(os.path.join(export_directory, export_path), arcname=export_path)
 
 
-def get_export_dataset_filename(name: str, ext: str, hid: int, conversion_key: Optional[str]) -> str:
+def get_export_dataset_filename(name: str, ext: str, encoded_id: str, conversion_key: Optional[str]) -> str:
     """
     Builds a filename for a dataset using its name an extension.
     """
     base = "".join(c in FILENAME_VALID_CHARS and c or "_" for c in name)
     if not conversion_key:
-        return f"{base}_{hid}.{ext}"
+        return f"{base}_{encoded_id}.{ext}"
     else:
-        return f"{base}_{hid}_conversion_{conversion_key}.{ext}"
+        return f"{base}_{encoded_id}_conversion_{conversion_key}.{ext}"
 
 
-def get_export_dataset_extra_files_dir_name(name: str, ext: str, hid: int, conversion_key: Optional[str]) -> str:
+def get_export_dataset_extra_files_dir_name(name: str, ext: str, encoded_id: str, conversion_key: Optional[str]) -> str:
     if not conversion_key:
-        return f"extra_files_path_{hid}"
+        return f"extra_files_path_{encoded_id}"
     else:
-        return f"extra_files_path_{hid}_conversion_{conversion_key}"
+        return f"extra_files_path_{encoded_id}_conversion_{conversion_key}"
 
 
 def imported_store_for_metadata(

--- a/test/unit/data/model/test_model_store.py
+++ b/test/unit/data/model/test_model_store.py
@@ -545,6 +545,21 @@ def validate_invocation_collection_crate_directory(crate_directory):
         assert dataset in root["hasPart"]
 
 
+def test_export_history_with_missing_hid():
+    # The dataset's hid was used to compose the file name during the export but it
+    # can be missing sometimes. We now use the dataset's encoded id instead.
+    app = _mock_app()
+    u, history, d1, d2, j = _setup_simple_cat_job(app)
+
+    # Remove hid from d1
+    d1.hid = None
+    app.commit()
+
+    temp_directory = mkdtemp()
+    with store.DirectoryModelExportStore(temp_directory, app=app, export_files="copy") as export_store:
+        export_store.export_history(history)
+
+
 def test_export_history_to_ro_crate(tmp_path):
     app = _mock_app()
     u, history, d1, d2, j = _setup_simple_cat_job(app)

